lib/ansible/modules/cloud/docker/docker_stack.py

```diff
@@ -0,0 +1,226 @@
+    returned: on change
+    type: dict
+'''
+
+EXAMPLES = '''
+-   name: deploy 'stack1' stack from file
+    docker_stack:
+        state: present
+        name: stack1
+        compose_file: /opt/stack.compose
+-   name: deploy 'stack2' from yaml
+    docker_stack:
+        state: present
+        name: stack2
+        compose_yaml: |
+            version: '3'
+            services:
+                web:
+                    image: nginx
+                    ports:
+                    -   "80:80"
+-   name: deprovision 'stack1'
+    docker_stack:
+        state: absent
+'''
+
+from ansible.module_utils.basic import AnsibleModule, os
+from tempfile import mkstemp
+import json
+from jsondiff import diff as json_diff
+
+
+def docker_stack_services(module, stack_name):
+    rc, out, err = module.run_command(["docker",
+                                       "stack",
+                                       "services",
+                                       stack_name,
+                                       "--format",
+                                       "{{.Name}}"])
+    if out != ("Nothing found in stack %s\n" % stack_name):
+        return(out.strip().split('\n'))
+    return []
+
+
+def docker_service_inspect(module, service_name):
+    rc, out, err = module.run_command(["docker",
+                                       "service",
+                                       "inspect",
+                                       service_name])
+    if rc != 0:
+        return None
+    else:
+        ret = json.loads(out)[0]['Spec']
+        return(ret)
+
+
+def docker_stack_deploy(module, stack_name, compose_file):
+    command = ["docker", "stack", "deploy"]
+    if module.params["prune"]:
+        command += ["--prune"]
+    command += ["--compose-file",
+                compose_file,
+                stack_name]
+    return module.run_command(command)
+
+
+def docker_stack_inspect(module, stack_name):
+    ret = {}
+    for service_name in docker_stack_services(module, stack_name):
+        ret[service_name] = docker_service_inspect(module, service_name)
+    return ret
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'name': dict(required=True),
+            'compose_yaml': dict(),
+            'compose_file': dict(),
+            'prune': dict(default=False, type='bool'),
+            'state': dict(default='present', choices=['present', 'absent'])
+        },
+        supports_check_mode=False
+    )
+
+    state = module.params['state']
+    compose_yaml = module.params['compose_yaml']
+    compose_file = module.params['compose_file']
+    name = module.params['name']
+
+    if state == 'present':
+        if compose_yaml and compose_file:
+            module.fail_json(msg="both compose_file and compose_yaml " +
+                                 "parameters given")
+        elif compose_yaml:
+            fd, compose_file = mkstemp()
+            stack_file = open(compose_file, 'w')
```

Comment:
Ok, I also enclosed the whole part of code which uses the temporary file in a try-finally block, to ensure removal of the temporary file